micrometer-core/src/main/java/io/micrometer/core/instrument/push/PushMeterRegistry.java

```diff
@@ -82,7 +82,14 @@ public void start(ThreadFactory threadFactory) {
     public void stop() {
         if (scheduledExecutorService != null) {
             scheduledExecutorService.shutdown();
-            scheduledExecutorService = null;
+            try {
+                scheduledExecutorService.awaitTermination(config.shutdownTimeout().toMillis(), TimeUnit.MILLISECONDS);
+            } catch (InterruptedException ie) {
+                scheduledExecutorService.shutdownNow();
+                Thread.currentThread().interrupt();
```

Comment:
I'm thinking if doing this is a good idea rather than just log a `WARN` event and carry on (assuming this happens at JVM shutdown).
What do you think?

Reply:
@jonatan-ivanov The backstory on how I ran into this was that I have an AWS Lambda container running on Micronaut with Micrometer for metrics. Since the container is paused between Lambda invocations, you can't get metrics pushed at a reliable interval because the push interval goes from 60 seconds of wall clock time to 60 seconds of invocation time, which can be awhile for infrequently invoked lambdas.
In order to push the metrics, I'm actually invoking [`PushMeterRegistry#publish`](https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/push/PushMeterRegistry.java#L45 )after every Lambda invocation (via reflection since it's not public - and maybe that's another PR). 
My other idea was to start/stop the `PushMeterRegistry` around every Lambda invocation, but `PushMeterRegistry #stop()` didn't block because of the missing `awaitTermination()` call, this didn't behave as I expected, thus this PR.

Reply:
Also, calling `shutdown` and then `awaitTermination` is the proper way to nicely shutdown an `ExecutorService`...

Reply:
I was only talking about these two lines:
```java
scheduledExecutorService.shutdownNow();
Thread.currentThread().interrupt();
```
As far as I remember, calling just `shutdown` is a proper way to nicely shutdown an `ExecutorService` **if** you don't want to block on it or you don't care about the results.

But if you do, then this is the proper way of nicely shutting down an `ExecutorService`, please see the [javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html):

```java
 void shutdownAndAwaitTermination(ExecutorService pool) {
   pool.shutdown(); // Disable new tasks from being submitted
   try {
     // Wait a while for existing tasks to terminate
     if (!pool.awaitTermination(60, TimeUnit.SECONDS)) {
       pool.shutdownNow(); // Cancel currently executing tasks
       // Wait a while for tasks to respond to being cancelled
       if (!pool.awaitTermination(60, TimeUnit.SECONDS))
           System.err.println("Pool did not terminate");
     }
   } catch (InterruptedException ie) {
     // (Re-)Cancel if current thread also interrupted
     pool.shutdownNow();
     // Preserve interrupt status
     Thread.currentThread().interrupt();
   }
 }
```

What I meant to say with my original comment is instead of calling `shutdownNow` and `interrupt` the `Thread` again, wouldn't it be simpler just to log a `WARN` event that says `"Waiting for the threadpool to shut down timed out."`? Assuming this happens at JVM shutdown, this may (or may not) give your tasks a little bit of an extra time to finish their job instead of cancelling them immediately.

Reply:
@shakuzen Do you happen to know if `stop` should be fire-and-forget or can we block on it for a while?

Reply:
@jonatan-ivanov Sorry, I didn't fully understand your prior comment, but do now.